datasets/wikisource/wikisource.py

```diff
@@ -0,0 +1,236 @@
+            if ".xml" not in fname:
+                continue
+            total_bytes += info["size"]
+            xml_urls.append(_base_url(lang) + fname)
+
+        # Use dictionary since testing mock always returns the same result.
+        downloaded_files = dl_manager.download({"xml": xml_urls})
+        if not pipeline.is_local():
+            downloaded_files = dl_manager.ship_files_with_pipeline(downloaded_files, pipeline)
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN, gen_kwargs={"filepaths": downloaded_files["xml"], "language": lang}
+            )
+        ]
+
+    def _build_pcollection(self, pipeline, filepaths, language):
+        """Build PCollection of examples in the raw (text) form."""
+        import apache_beam as beam
+        import mwparserfromhell
+
+        def _extract_content(filepath):
+            """Extracts article content from a single WikiMedia XML file."""
+            logger.info("generating examples from = %s", filepath)
+            with beam.io.filesystems.FileSystems.open(filepath) as f:
+                f = bz2.BZ2File(filename=f)
+                # Workaround due to: https://github.com/tensorflow/tensorflow/issues/33563
+                utf_f = codecs.getreader("utf-8")(f)
+                context = etree.iterparse(utf_f, events=("end",))
+                for unused_event, elem in context:
+                    if not elem.tag.endswith("page"):
+                        continue
+                    namespace = elem.tag[:-4]
+                    title = elem.find(f"./{namespace}title").text
+                    ns = elem.find(f"./{namespace}ns").text
+                    id_ = elem.find(f"./{namespace}id").text
+
+                    # Filter pages that are not in the "main" namespace.
+                    if ns != "0":
+                        elem.clear()
+                        continue
+
+                    raw_content = elem.find(f"./{namespace}revision/{namespace}text").text
+                    elem.clear()
+
+                    # Filter redirects.
+                    if raw_content is None or raw_content.lower().startswith("#redirect"):
+                        beam.metrics.Metrics.counter(language, "filtered-redirects").inc()
+                        continue
+
+                    beam.metrics.Metrics.counter(language, "extracted-examples").inc()
+                    yield (id_, title, raw_content)
+
+        def _clean_content(inputs):
+            """Cleans raw wikicode to extract text."""
+            id_, title, raw_content = inputs
+            try:
+                text = _parse_and_clean_wikicode(raw_content, parser=mwparserfromhell)
+            except (mwparserfromhell.parser.ParserError) as e:
+                beam.metrics.Metrics.counter(language, "parser-error").inc()
+                logger.error("mwparserfromhell ParseError: %s", e)
+                return
+
+            if not text:
```

Comment:
Given that it's not uncommon for Wikisource pages to have a small amount of metadata (e.g., a few categories) but not contain the transcribed text of the document yet, we probably want to also add a basic length filter. In my exploration, a minimum of 100 characters seems to be a good trade-off between removing many useless documents while retaining those that do have content. So: `if not text or len(text) < 100:` (and obviously could move the 100 up to a config variable to make it more salient of a choice.

Reply:
Done.